src/metabase/query_processor/middleware/upgrade_field_literals.clj

```diff
@@ -1,36 +1,113 @@
 (ns metabase.query-processor.middleware.upgrade-field-literals
-  (:require [clojure.walk :as walk]
+  (:require [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [clojure.walk :as walk]
+            [metabase.config :as config]
             [metabase.mbql.util :as mbql.u]
-            [metabase.util :as u]))
+            [metabase.query-processor.middleware.resolve-fields :as resolve-fields]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.util :as u]
+            [metabase.util.i18n :refer [trs]]))
+
+(defn- has-a-native-source-query-at-some-level? [{:keys [source-query]}]
+  (or (:native source-query)
+      (when source-query
+        (has-a-native-source-query-at-some-level? source-query))))
+
+(defn- warn-once
+  "Log only one warning per QP run (regardless of message)."
+  [message]
+  ;; Make sure QP store is available since we use caching below (it may not be in some unit tests)
+  (when (qp.store/initialized?)
+    ;; by caching the block below, the warning will only get trigger a maximum of one time per query run. We don't need
+    ;; to blow up the logs with a million warnings.
+    (qp.store/cached ::bad-clause-warning
+      (log/warn (u/colorize :red message)))))
+
+(defn- fix-clause [{:keys [inner-query source-aliases field-name->field]} [_ field-name options :as field-clause]]
+  ;; attempt to find a corresponding Field ref from the source metadata.
+  (let [field-ref (:field_ref (get field-name->field field-name))]
+    (cond
+      field-ref
+      (mbql.u/match-one field-ref
+        ;; If the matching Field ref is an integer `:field` clause then replace it with the corrected clause and log
+        ;; a developer-facing warning. Things will still work and this should be fixed on the FE, but we don't need to
+        ;; blow up prod logs
+        [:field (id :guard integer?) new-options]
+        (u/prog1 [:field id (merge new-options (dissoc options :base-type))]
+          (when (and (not config/is-prod?)
+                     (not (has-a-native-source-query-at-some-level? inner-query)))
+            ;; don't i18n this because it's developer-facing only.
+            (warn-once
+             (str "Warning: query is using a [:field <string> ...] clause to refer to a Field in an MBQL source query."
```

Comment:
this warning says it is using `[:field <string> ...]` but the match clause is `[:field (id :guard integer?) new-options]`. They seem incongruous.

Reply:
ah i missed it. I see now.